src/util/Backtrace.cpp

```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "util/Backtrace.h"
+#include "config.h"
+#include <cstdio>
+#include <cstdlib>
+#include <stdexcept>
+
+namespace
+{
+
+// The abi::__cxa_demangle call in printCurrentBacktrace below uses a buffer
+// that must be a _malloc()_ memory-region, so that it can realloc() if the
+// region is too small. To avoid calling malloc() in the middle of unwinding
+// (which might mess with already-corrupt state) we pre-malloc() a fairly large
+// buffer using BacktraceManager, with hope that it is enough for any symbol we
+// encounter.
+//
+// The BacktraceManager class is idempotent and printCurrentBacktrace sets one
+// up locally itself, in case nobody has done so yet. This minimizes the number
+// of malloc/free calls and ensures we don't leak (except in a memory-allocation
+// error path below where .. we really don't know what the right recovery is).
+char* demangleBuf = nullptr;
```

Comment:
I wish! Unfortunately as mentioned in the comments, abi::__cxa_demangle literally _requires_ that its buffer is a malloc-region (it can call realloc if it runs out of room).